### PR TITLE
Access individual `map` iteration results using `from`

### PR DESCRIPTION
**Pull request by juniper-shopify**
Building on the architectural pattern established down-stack for accessing output from a specific, individual iteration of a `repeat` loop, you can now access a specific iteration of a `map` invocation as well.

Note that the semantics here are slightly different. `map` is run on an iterable of some known size, N, and it will always produce the results of N iterations. However, because `break!` can be issued inside the map executor, some of those iterations that had not started yet will be `nil` in the map result. Attempting to access those iterations directly will raise an `IndexError`.

To help make this uncertainty easier to handle, `Map::Output` get an `iteration?` method that will check if the requested iteration has output, if you know you're working with a map that might not always complete (i.e. contains a conditional `break!`). I expect this usage will be quite uncommon for `map`, so I don't think additional shorthand is required at this time.

### CODE REVIEW

**Review by mathiusj — APPROVED** [review]
curious if we can have any functional tests for the errors that are raised, but non-blocking!

**Comment by juniper-shopify** [comment]
### Merge activity

* **Nov 27, 3:28 AM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/Shopify/roast/588).
* **Nov 27, 3:29 AM UTC**: @juniper-shopify merged this pull request with [Graphite](https://app.graphite.com/github/pr/Shopify/roast/588).

File: lib/roast/dsl/system_cogs/map.rb [hunk 1]
```diff
@@ -5,6 +5,10 @@ module Roast
   module DSL
     module SystemCogs
       class Map < SystemCog
+        class MapOutputAccessError < Roast::Error; end
+
+        class MapIterationDidNotRunError < MapOutputAccessError; end
```
[the remark this answers is not in the document]
Comment (juniper-shopify) [reply lib/roast/dsl/system_cogs/map.rb]: Ultimately, yes, we should do this. So far, because the functional tests have also served as usage examples, we've been reluctant to explicitly include failure scenarios, because we don't want the example workflows to fail if a user just runs them at the command line (and putting a `rescue` statement in the workflow itself is clunky and makes for a bad example, since no one should have to do that sort of thing normally).

You're pointing out a systemic / structural we should fix soon -- we're overloading the purpose of the examples in the `dsl/` directory. Initially, we started writing them as examples of the new syntax we were developing, just to show each other what it looked like, and see that it worked. Then, we started breaking things, so we added a quick-and-dirty functional test suite that just ran all of our example workflows. This saved us from making major, obvious regressions, but as a result we didn't have a good way to test failure modes (or much incentive to do so at that point). We now need to separate out these concerns into separate things -- runnable examples that always work and show how to write Roast workflows 'correctly' and functional tests, some of which intentionally fail to run and which test valid but weird ways of writing workflows. (We also need to write unit tests, but that's a whole separate thing).

I've started writing a tutorial with a set of runnable example workflows that I think will do a better job of teaching a new user the Roast system from the ground up, in a way that's easier to follow than just poking at the examples we have now. This starts to separate example workflows targeted at the user from functional tests exercising internal edge cases (we should still run the example workflows in the functional test suite, although we'll need VCR set up for that)
